File: regexp.py
# TODO: add comments to this file mirroring Regexp.hs
class Re(object):
    def match(self, str, idx, cont):
        raise NotImplementedError()

    def matches(self, str):
        return self.match(str, 0, lambda i: i == len(str))

class ReChar(Re):
    # ReChar(c) represents the character literal c.
    def __init__(self, char): self.char = char
    def match(self, str, idx, cont):
        # If the character at idx is the one we're looking for, we match!
        if len(str) > idx and str[idx] == self.char:
            # If we match, we call our continuation, to check that the rest of
            # the string matches
            return cont(idx+1)
        return False            # we didn't match

class ReSeq(Re):
    # ReSeq(a, b) represents /ab/
    def __init__(self, first, second): self.first, self.second = first, second
    def match(self, str, idx, cont):
        # first try matching self.first
        return self.first.match(str, idx,
                                # giving a modified continuation that checks
                                # whether self.second matches
                                lambda idx: self.second.match(str, idx, cont))

class ReOr(Re):
    # ReOr(a,b) represents /a|b/
    def __init__(self, first, second): self.first, self.second = first, second
    def match(self, str, idx, cont):
        return (self.first.match(str, idx, cont) # try matching self.first
                or self.second.match(str, idx, cont) # else try self.second
        )
        # Note how we might call our continuation twice - backtracking!

class ReStar(Re):
    # ReStar(a) represents /a*/
    def __init__(self, inner): self.inner = inner
    def match(self, str, idx, cont):
        return (
            # the one-or-more case, like /aa*/
            self.inner.match(str, idx,
                                 lambda idx: self.match(str, idx, cont))
            # the empty case, like //
            or cont(idx))

File: Regexp.hs
module Regexp where

data Re
    -- character literals
    = Lit Char
    -- (Seq a b) is concatenation /ab/
    | Seq Re Re
    -- (Or a b) is choice /a|b/
    | Or Re Re
    -- (Star a) is iteration /a*/
    | Star Re
      deriving (Show, Eq)

match :: Re -> String -> (String -> Bool) -> Bool
match (Lit c) [] k = False      -- a literal doesn't match the empty string!
match (Lit c) (x:xs) k =
    if c == x                   -- if the first character matches,
    then k xs                   -- try matching the rest of the string
    else False                  -- (if it didn't, we're screwed)

match (Seq a b) s k =
    match a s                       -- match a...
          (\rest -> match b rest k) -- then match b
-- Note how we modified our continuation - we added more work to do!

match (Or a b) s k =
    match a s k                 -- try matching a...
    || match b s k              -- else try b
-- Note how we might call our continuation twice - backtracking!

match (Star a) s k =
    -- one or more case, /aa*/
    match a s (\rest -> match (Star a) rest k)
    -- empty case, //
    || k s

matches :: Re -> String -> Bool
matches a s = match a s null
-- We pass null because we want to match the *whole* string - so the "work left
-- to do" is to check that there's nothing left in the string, which is what
-- null does.

-- A more concise implementation.
match2 :: Re -> String -> (String -> Bool) -> Bool
match2 (Lit c) [] k = False
match2 (Lit c) (x:xs) k = c == x && k xs
match2 (Seq a b) s k = match2 a s (\rest -> match2 b rest k)
match2 (Or a b) s k = match2 a s k || match2 b s k
match2 (Star a) s k = match2 a s (\rest -> match2 (Star a) rest k) || k s
